File: mapping-engine-streamsets-plugin/src/main/java/com/ebsco/entarch/mappings/streamsets/XmlToJsonCanonicalProcessor.java
package com.ebsco.entarch.mappings.streamsets;

import com.ebsco.platform.shared.mappingsengine.config.MappingsEngineJsonConfig;
import com.ebsco.platform.shared.mappingsengine.config.Xml2JsonConfig;
import com.ebsco.platform.shared.mappingsengine.core.MappingsEngine;
import com.ebsco.platform.shared.mappingsengine.xml.XmlToRecordParser;
import com.ebsco.platform.shared.mappingsengine.xml.XmlToRecordParserConfig;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.jayway.jsonpath.spi.json.JacksonJsonProvider;
import com.streamsets.pipeline.api.*;
import com.streamsets.pipeline.api.base.OnRecordErrorException;
import com.streamsets.pipeline.api.base.SingleLaneRecordProcessor;
import lombok.Getter;

import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

@StageDef(version = 1, label = "XML2JSON Canonical Processor", icon = "default.png", onlineHelpRefUrl = "")
@ConfigGroups(Groups.class)
@GenerateResourceBundle
public class XmlToJsonCanonicalProcessor extends SingleLaneRecordProcessor {

    private static final String EMPTY_CONFIG = "{\n" +
            "  \"metadata\": {\n" +
            "    \"id\": \"empty\",\n" +
            "    \"version\": \"1.0.0\",\n" +
            "    \"primaryKey\": \"$.id\"\n" +
            "  },\n" +
            "  \"transforms\": [\n" +
            "  ],\n" +
            "  \"configuration\": {\n" +
            "    \"xml2json\": {\n" +
            "    }\n" +
            "  }\n" +
            "}";

    @ConfigDef(required = true, type = ConfigDef.Type.TEXT, defaultValue = EMPTY_CONFIG,
            label = "Mappings JSON", displayPosition = 10, group = "MAPPINGS",
            mode = ConfigDef.Mode.JSON,
            lines = 15,
            description = "Mappings instruction set as JSON")
    public String mappingInstructions = null;

    @ConfigDef(required = true, type = ConfigDef.Type.MODEL,
            defaultValue = "/fileRef",
            label = "Input Field", displayPosition = 6, group = "MAPPINGS",
            description = "Input Field, can be fileRef or string field containing content")
    @FieldSelectorModel(singleValued = true)
    public String rawXmlField = null; // TODO: slightly misnamed but breaks existing flows to rename

    @ConfigDef(required = true, type = ConfigDef.Type.MODEL,
               defaultValue = "XML",
               label = "Input Type", displayPosition = 5, group = "MAPPINGS",
               description = "Input format (XML, JSON, ...)")
    @ValueChooserModel(InputTypesChooser.class)
    public InputTypes rawInputType = InputTypes.XML;

    @ConfigDef(required = true, type = ConfigDef.Type.MODEL,
            defaultValue = "/json",
            label = "JSON output Field", displayPosition = 8, group = "MAPPINGS",
            description = "Output field to place the text version of the resulting JSON")
    @FieldSelectorModel(singleValued = true)
    public String outJsonField = null;

    @Getter(lazy = true)
    private final MappingsEngineJsonConfig cfgJson = MappingsEngineJsonConfig.fromJson(mappingInstructions);

    @Getter(lazy = true)
    private final Xml2JsonConfig xml2JsonConfig = getCfgJson().getConfiguration().getXml2json();

    @Getter(lazy = true)
    private final XmlToRecordParser parser = buildParser();

    @Getter(lazy = true)
    private final MappingsEngine mappings = MappingsEngine.builder()
            .transforms(getCfgJson().getTransforms())
            .jsonProvider(getParser().getJson())
            .build();

    private final ObjectMapper jsonMapper = new ObjectMapper();


    @Override
    public List<ConfigIssue> init(Info info, Processor.Context context) {
        List<ConfigIssue> issues = super.init(info, context);
        if (mappingInstructions.equals("invalidValue")) {
            issues.add(context.createConfigIssue(Groups.MAPPINGS.getLabel(), "mappingInstructions",
                    Errors.EBSCO_INVALID_CONFIG, "mappings mappingInstructions JSON is invalid")); // TODO: add errors from mapping system
        }
        if (rawXmlField.equals("invalidValue")) {
            issues.add(context.createConfigIssue(Groups.MAPPINGS.getLabel(), "rawXmlField",
                    Errors.EBSCO_INVALID_CONFIG, "field name for XML input is invalid")); // TODO: add errors from mapping system
        }
        return issues;
    }

    @Override
    public void destroy() {
        super.destroy();
    }

    @Override
    protected void process(Record record, SingleLaneBatchMaker batchMaker) throws StageException {
        applyMappings(record, rawXmlField);
        batchMaker.addRecord(record);
    }

    private void applyMappings(Record record, String inputFieldName) throws OnRecordErrorException {
        try {
            String inputFieldNameFixed = fixInputFieldName(inputFieldName);
            Field inputField = record.get(inputFieldNameFixed);
            try(InputStream is = getInputStream(inputField)) {
                final Map<String, Map<String, Object>> jsonObject;
                if (rawInputType == InputTypes.XML) {
                    XmlToRecordParser.Result parsed = getParser().parse(is);
                    jsonObject = new HashMap<>();
                    jsonObject.put(parsed.getName(), (Map<String, Object>) parsed.getJsonNode());
                } else if (rawInputType == InputTypes.JSON) {
                    jsonObject = jsonMapper.readValue(is, Map.class);
                } else {
                    throw new IllegalArgumentException("Input Type " + rawInputType.name() + " is not supported.");
                }
                getMappings().processDocument(jsonObject);

                String prettyJson = jsonMapper.writerWithDefaultPrettyPrinter().writeValueAsString(jsonObject);
                Field field = Field.create(prettyJson);
                if (outJsonField == null) {
                    record.set("/json", field);
                } else {
                    record.set(outJsonField, field);
                }
            }
        } catch (Exception e) {
            throw new OnRecordErrorException(record, Errors.EBSCO_RECORD_ERROR, e.getMessage());
        }
    }

    private InputStream getInputStream(Field inputField) throws IOException {
        InputStream inputStream;
        if (inputField.getType() == Field.Type.STRING) {
            inputStream =  new ByteArrayInputStream(inputField.getValueAsString().getBytes());
        } else {
            inputStream = inputField.getValueAsFileRef().createInputStream(getContext(), InputStream.class);
        }
        return inputStream;
    }

    private String fixInputFieldName(String inputFieldName) {
        if (inputFieldName != null && !inputFieldName.trim().isEmpty() && inputFieldName.startsWith("/")) {
            return inputFieldName;
        } else {
            return "/fileRef";
        }
    }

    private XmlToRecordParser buildParser() {
        Xml2JsonConfig xml2JsonConfig = getXml2JsonConfig();
        XmlToRecordParserConfig xmlToRecordParserConfig = XmlToRecordParserConfig.builder()
                .preserveNestedTextElements_ByXPath(xml2JsonConfig.getEmbedLiteralXmlAtPaths())
                .preserveNestedTextElements_AutoDetect(xml2JsonConfig.isAutoDetectMixedContent())
                .preserveNestedTextElements_UnhandledResultInError(xml2JsonConfig.isUnhandledMixedContentIsError())
                .forceSingleValueNodes_ByXPath(xml2JsonConfig.getForceSingleValueElementAtPaths())
                .forceElevateTextNode_ByXPath(xml2JsonConfig.getForceElevateTextNodesAtPaths())
                .forceElevateTextNodesAreSingleValued(xml2JsonConfig.isForceElevateTextNodesAsSingleValue())
                .textNodeName(xml2JsonConfig.getTextNodeName())
                .attributeNodePrefix(xml2JsonConfig.getAttributeNodePrefix())
                .attributePrefixesToKeep(xml2JsonConfig.getPreserveAttributePrefixes())
                .jsonProvider(new JacksonJsonProvider())
                .build();
        return new XmlToRecordParser(xmlToRecordParserConfig);
    }
}

File: mapping-engine-streamsets-plugin/src/main/kotlin/com/ebsco/entarch/mappings/streamsets/XmlToJsonCanonicalProcessor.kt
package com.ebsco.entarch.mappings.streamsets

import com.ebsco.platform.shared.mappingsengine.config.MappingsEngineJsonConfig
import com.ebsco.platform.shared.mappingsengine.core.DefaultTransformers
import com.ebsco.platform.shared.mappingsengine.core.MappingsEngine
import com.ebsco.platform.shared.mappingsengine.xml.XmlToRecordParser
import com.ebsco.platform.shared.mappingsengine.xml.XmlToRecordParserConfig
import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import com.jayway.jsonpath.spi.json.JacksonJsonProvider
import com.streamsets.pipeline.api.*
import com.streamsets.pipeline.api.base.OnRecordErrorException
import com.streamsets.pipeline.api.base.SingleLaneRecordProcessor
import org.slf4j.LoggerFactory
import java.io.InputStream

@StageDef(version = 1, label = "XML2JSON Canonical Processor", description = "", icon = "default.png", onlineHelpRefUrl = "")
@ConfigGroups(Groups::class)
@GenerateResourceBundle
class XmlToJsonCanonicalProcessor : SingleLaneRecordProcessor() {
    val LOG = LoggerFactory.getLogger(XmlToJsonCanonicalProcessor::class.java)
    @ConfigDef(required = true, type = ConfigDef.Type.TEXT, defaultValue = emptyConfigJson,
            label = "Mappings JSON", displayPosition = 10, group = MAPPINGS_GROUP,
            description = "Mappings instruction set as JSON, either Simple or SLS formats")
    @JvmField
    var mappingInstructions: String? = null

    @ConfigDef(required = true, type = ConfigDef.Type.MODEL,
            defaultValue = """/fileRef""",
            label = "Raw XML Field", displayPosition = 5, group = MAPPINGS_GROUP,
            description = "Raw XML Field, can be fileRef or string field containing XML")
    @FieldSelectorModel(singleValued = true)
    @JvmField
    var rawXmlField: String? = null

    @ConfigDef(required = true, type = ConfigDef.Type.MODEL,
            defaultValue = """/json""",
            label = "JSON output Field", displayPosition = 8, group = MAPPINGS_GROUP,
            description = "Output field to place the text version of the resulting JSON")
    @FieldSelectorModel(singleValued = true)
    @JvmField
    var outJsonField: String? = null


    private val cfgJson by lazy {
        MappingsEngineJsonConfig.fromJson(mappingInstructions!!)
    }

    private val xml2jsonCfg by lazy { cfgJson.configuration.xml2json }

    private val parser by lazy {
        XmlToRecordParser(XmlToRecordParserConfig(
                preserveNestedTextElements_ByXPath = xml2jsonCfg.embedLiteralXmlAtPaths,
                preserveNestedTextElements_AutoDetect = xml2jsonCfg.isAutoDetectMixedContent,
                preserveNestedTextElements_UnhandledResultInError = xml2jsonCfg.isUnhandledMixedContentIsError,
                forceSingleValueNodes_ByXPath = xml2jsonCfg.forceSingleValueElementAtPaths,
                forceElevateTextNode_ByXPath = xml2jsonCfg.forceElevateTextNodesAtPaths,
                forceElevateTextNodesAreSingleValued = xml2jsonCfg.isForceElevateTextNodesAsSingleValue,
                textNodeName = xml2jsonCfg.textNodeName,
                attributeNodePrefix = xml2jsonCfg.attributeNodePrefix,
                attributePrefixesToKeep = xml2jsonCfg.preserveAttributePrefixes.toSet(),
                jsonProvider = JacksonJsonProvider()
        ))
    }

    private val mappings by lazy { MappingsEngine(cfgJson.transforms, DefaultTransformers.TRANFORMERS, parser.config.jsonProvider) }

    private val jsonMapper = jacksonObjectMapper()

    override fun init(): MutableList<Stage.ConfigIssue> {
        return super.init().also { issues ->
            if (mappingInstructions == "invalidValue") {
                issues.add(context.createConfigIssue(Groups.Mappings.groupName, XmlToJsonCanonicalProcessor::mappingInstructions.name,
                        Errors.EBSCO_INVALID_CONFIG, "mappings mappingInstructions JSON is invalid")) // TODO: add errors from mapping system
            }
            if (rawXmlField == "invalidValue") {
                issues.add(context.createConfigIssue(Groups.Mappings.groupName, XmlToJsonCanonicalProcessor::rawXmlField.name,
                        Errors.EBSCO_INVALID_CONFIG, "field name for XML input is invalid")) // TODO: add errors from mapping system
            }
        }
    }

    override fun destroy() {
        super.destroy()
    }

    fun String.mustStartWith(s: String) = if (this.startsWith(s)) this else s + this

    fun fixXmlToBetterJson(record: Record, inputFieldName: String = rawXmlField!!): Record {
        try {
            val inputFieldNameFixed = inputFieldName.takeIf { it.isNotBlank() }?.mustStartWith("/")
                    ?: "/fileRef"
            val inputField = record.get(inputFieldNameFixed)

            val inputStream = if (inputField.type == Field.Type.STRING) {
                inputField.valueAsString.toByteArray().inputStream()
            } else {
                inputField.valueAsFileRef.createInputStream(context, InputStream::class.java)
            }

            val (xmlField, xmlData) = inputStream.use { parser.parse(it) }

            @Suppress("UNCHECKED_CAST")
            val jsonObject = mapOf(xmlField to xmlData as Map<String, Any>)

            mappings.processDocument(jsonObject)

            // we go direct to a JSON field, which can be parsed by streamsets if desired to be operated upon further.

            val json = jsonMapper.writerWithDefaultPrettyPrinter().writeValueAsString(jsonObject)

            record.set(outJsonField ?: "/json", Field.create(json))

            return record
        } catch (ex: Throwable) {
            throw OnRecordErrorException(record, Errors.EBSCO_RECORD_ERROR, ex.message)
        }
    }

    override fun process(record: Record, batchMaker: SingleLaneBatchMaker) {
        fixXmlToBetterJson(record)
        batchMaker.addRecord(record)
    }


}